### PR TITLE
Bug fix for the variable projection algorithm of BOPDMD 

### PR DESCRIPTION
**Pull request by dsj976**
There is a small bug in the variable projection algorithm implemented in the `BOPDMD` class. This bug was causing the algorithm to fail to converge for certain "complex" signals (e.g. composed of three sine waves).

This is the [relevant line](https://github.com/duqbo/optdmd/blob/eff57045ab4085ae2b6a1a2decb583deeb1be488/src/varpro2.m#L272) in the original Matlab code.

I have made a [repo](https://github.com/dsj976/test-opt-dmd) were you can test these changes on a pre-generated dataset. Read the `README.md` for instructions.

### CODE REVIEW

File: pydmd/bopdmd.py [hunk 1]
```diff
@@ -624,8 +624,8 @@ def compute_B(alpha):
             q_out, djac_out, j_pvt = qr(
                 djac_matrix, mode="economic", pivoting=True
             )
-            ij_pvt = np.arange(IA)
-            ij_pvt = ij_pvt[j_pvt]
+            ij_pvt = np.zeros(IA).astype(int)
+            ij_pvt[j_pvt] = np.arange(IA).astype(int)
```
Comment (fandreuz) [inline pydmd/bopdmd.py:628]:
```suggestion
            ij_pvt = np.zeros(IA, dtype=int)
            ij_pvt[j_pvt] = np.arange(IA, dtype=int)
```

**Comment by sichinaga** [comment]
Thank you @dsj976 for the PR and for catching this!

The changes to `bopdmd.py` look good to me, though I'm not sure how necessary the `.gitignore` changes are.

Will be awaiting more input from the rest of the team!

**Comment by fandreuz** [comment]
Thanks @dsj976 ! This is good to go for me

**Comment by klapo** [comment]
Hi @dsj976, could you describe how BOPDMD failed for the 3 sine wave case? I suspect this fix does not quite behave how we suspect it should.

**Comment by dsj976** [comment]
Hi @klapo. The problem was related to differences in the generation of the `ij_pvt` index vector in the Python and Matlab (where it is referred to as `ijpvt`) versions of optimized DMD. In the Python version, this lead to convergence issues in the variable projection algorithm, so the output of `BOPDMD` was essentially the same as applying exact DMD directly (because exact DMD is used to initialize `alpha`). After applying the fix in this PR, the convergence rate of the variable projection algorithm was exactly the same in the Python and Matlab versions of the code. Details can be found in the repo I linked in the PR description.

**Comment by klapo** [comment]
Hi @dsj976, first of all a huge thanks for so closely documenting your work. I _think_ I have the source of where things go off the rails.

- Fitting without eigenvalue constraints: your fix works beautifully. 
- Fitting with the "imag" eigenvalue constraints (and based on my accidental testing potentially other eigenvalue constraints): your fix seems to degrade the DMD solutions.

Could you verify if that is the case and not just me making a mistake? The update I have coming down the line makes extensive use of eigenvalue constraints for bopdmd.

**Comment by dsj976** [comment]
Hi @klapo, thanks for the update. I will have a look at this and get back to you by the end of the week.

**Comment by dsj976** [comment]
Hi @klapo, just had a look at this and these are the results I get for Python (with my fix and without it) and the original Matlab code. The synthetic data on which I run opt DMD consists of three sinusoids of the following characteristics:

1. Sinusoid 1: amp=2, omega=0.5
1. Sinusoid 2: amp=1, omega=2.5
1. Sinusoid 3: amp=1, omega=5

## Results

### 1. Python; with my fix; no eigenvalue constraints

```
Step 1 Error 0.40238704984388624 Lambda 0.3333333333333333
Step 2 Error 0.4023772838022436 Lambda 42.666666666666664
Step 3 Error 0.40229937169705 Lambda 85.33333333333333
Step 4 Error 0.40228487788286 Lambda 85.33331732948632
Step 5 Error 0.4022807441044969 Lambda 85.1243209801669
Step 6 Error 0.40227881560753787 Lambda 40.115963466827395
Step 7 Error 0.4022746409018976 Lambda 13.371987822275798
Step 8 Error 0.40225775177482903 Lambda 4.457329274091933
Step 9 Error 0.4021890093317776 Lambda 1.4857764246973109
Step 10 Error 0.40192279977018947 Lambda 0.49525880823243695
Step 11 Error 0.40079670921862476 Lambda 0.1650862694108123
Step 12 Error 0.3969864204026219 Lambda 0.055028756470270765
Step 13 Error 0.38469061740287597 Lambda 0.018342918823423586
Step 14 Error 0.3585980845836368 Lambda 0.006114306274474528
Step 15 Error 0.2553425969801761 Lambda 0.002038102091491509
Step 16 Error 0.09779491564949808 Lambda 0.000679367363830503
Step 17 Error 0.07087876047095029 Lambda 0.000226455787943501
Step 18 Error 0.0708763035722032 Lambda 7.548526264783366e-05
Step 19 Error 0.07087630347840834 Lambda 2.5161754215944553e-05
Step 20 Error 0.07087630347840454 Lambda 8.38725140531485e-06
Stall detected: error reduced by less than 1e-12 times the error at the previous step. Iteration 20. Current error 0.07087630347840454. Consider increasing tol or decreasing eps_stall.

Eigenvalues:
[-1.16216065e-04+5.0002048j  -1.16216065e-04-5.0002048j
 -6.70200341e-05+0.50008228j -6.70200341e-05-0.50008228j
 -4.29170607e-04+2.50035487j -4.29170607e-04-2.50035487j]
Amplitudes:
[1.00351144 1.00351144 2.00721623 2.00721623 1.02340275 1.02340275]
```

### 2. Matlab; no eigenvalue constraints

```
step 1 err 4.903176e+02 lambda 3.333333e-01
step 2 err 4.902938e+02 lambda 4.266667e+01
step 3 err 4.901039e+02 lambda 8.533333e+01
step 4 err 4.900686e+02 lambda 8.533332e+01
step 5 err 4.900585e+02 lambda 8.512432e+01
step 6 err 4.900538e+02 lambda 4.011596e+01
step 7 err 4.900437e+02 lambda 1.337199e+01
step 8 err 4.900025e+02 lambda 4.457329e+00
step 9 err 4.898351e+02 lambda 1.485776e+00
step 10 err 4.891868e+02 lambda 4.952588e-01
step 11 err 4.864495e+02 lambda 1.650863e-01
step 12 err 4.772443e+02 lambda 5.502876e-02
step 13 err 4.481389e+02 lambda 1.834292e-02
step 14 err 3.894085e+02 lambda 6.114306e-03
step 15 err 1.974404e+02 lambda 2.038102e-03
step 16 err 2.896157e+01 lambda 6.793674e-04
step 17 err 1.521324e+01 lambda 2.264558e-04
step 18 err 1.521218e+01 lambda 7.548526e-05
step 19 err 1.521218e+01 lambda 2.516175e-05
step 20 err 1.521218e+01 lambda 8.387251e-06
stall detected: residual reduced by less than 1.000000e-12 times residual at previous step. iteration 20. current residual 1.521218e+01
Eigenvalues:
  -0.0001 - 5.0002i  -0.0001 + 5.0002i  -0.0001 - 0.5001i  -0.0001 + 0.5001i  -0.0004 - 2.5004i  -0.0004 + 2.5004i

Amplitudes:
    1.0035    1.0035    2.0072    2.0072    1.0234    1.0234
```

### 3. Python; with my fix; imaginary eigenvalue constraints

```
Step 1 Error 0.40238550423558117 Lambda 0.3333333333333333
Step 2 Error 0.40230700152685095 Lambda 85.33333333333333
Step 3 Error 0.4022972860216572 Lambda 91.36038368863468
Step 4 Error 0.4022948507673578 Lambda 127.99704511888032
Step 5 Error 0.4022944001283151 Lambda 233.03981104775428
Step 6 Error 0.40229439825387503 Lambda 465.8469927839862
Step 7 Error 0.402294397991672 Lambda 931.58277517761
Step 8 Error 0.4022943979315549 Lambda 1863.0357232075414
Step 9 Error 0.40229439791674626 Lambda 3725.8578399335534
Step 10 Error 0.4022943979130589 Lambda 7451.313397499242
Step 11 Error 0.40229439791213834 Lambda 14901.835466869097
Step 12 Error 0.4022943979119078 Lambda 29802.091733670175
Stall detected: error reduced by less than 1e-12 times the error at the previous step. Iteration 12. Current error 0.4022943979119078. Consider increasing tol or decreasing eps_stall.

Eigenvalues:
[ 0.+5.14981449j -0.-5.14981449j  0.+0.5001056j  -0.-0.5001056j
  0.+2.50033053j -0.-2.50033053j]
Amplitudes:
[0.21997626 0.21997626 2.00347753 2.00347753 1.00988586 1.00988586]
```

### 4. Matlab; imaginary eigenvalue constraints

```
step 1 err 4.903138e+02 lambda 3.333333e-01
step 2 err 4.901225e+02 lambda 8.533333e+01
step 3 err 4.900988e+02 lambda 8.532383e+01
step 4 err 4.900936e+02 lambda 8.532383e+01
step 5 err 4.900922e+02 lambda 8.532281e+01
step 6 err 4.900919e+02 lambda 8.532281e+01
step 7 err 4.900918e+02 lambda 8.532028e+01
step 8 err 4.900918e+02 lambda 8.532129e+01
step 9 err 4.900918e+02 lambda 8.527593e+01
step 10 err 4.900918e+02 lambda 8.549618e+01
step 11 err 4.900918e+02 lambda 8.364568e+01
step 12 err 4.900918e+02 lambda 1.243600e+02
step 13 err 4.900918e+02 lambda 4.145334e+01
step 14 err 4.900918e+02 lambda 1.658133e+02
step 15 err 4.900918e+02 lambda 5.527111e+01
step 16 err 4.900918e+02 lambda 2.210845e+02
step 17 err 4.900918e+02 lambda 7.369482e+01
step 18 err 4.900918e+02 lambda 2.947793e+02
step 19 err 4.900918e+02 lambda 9.825976e+01
step 20 err 4.900918e+02 lambda 3.930390e+02
step 21 err 4.900918e+02 lambda 1.310130e+02
step 22 err 4.900918e+02 lambda 5.240520e+02
step 23 err 4.900918e+02 lambda 1.746840e+02
step 24 err 4.900918e+02 lambda 6.987361e+02
step 25 err 4.900918e+02 lambda 2.329120e+02
step 26 err 4.900918e+02 lambda 9.316481e+02
step 27 err 4.900918e+02 lambda 3.105494e+02
step 28 err 4.900918e+02 lambda 1.242197e+03
step 29 err 4.900918e+02 lambda 4.140658e+02
step 30 err 4.900918e+02 lambda 1.656263e+03
failed to reach tolerance after maxiter = 30 iterations. current residual 4.900918e+02
Eigenvalues:
   0.0000 - 5.1498i   0.0000 + 5.1498i   0.0000 - 0.5001i   0.0000 + 0.5001i   0.0000 - 2.5003i   0.0000 + 2.5003i

Amplitudes:
    0.2200    0.2200    2.0035    2.0035    1.0099    1.0099
```

### 5. Python; without my fix; no eigenvalue constraints

```
Step 1 Error 0.4160103125293694 Lambda 0.9780279453349598
Failed to find appropriate step length at iteration 2. Current error 0.4160103125293694. Consider increasing maxlam or lamup.

Eigenvalues:
[ 0.00275844+5.13633318j  0.00275844-5.13633318j  0.00034539+0.49864003j
  0.00034539-0.49864003j -0.00468287+2.51365699j -0.00468287-2.51365699j]
Amplitudes:
[0.18541474 0.18541474 1.98107684 1.98107684 1.12080351 1.12080351]
```

### 6. Python; without my fix; with imaginary eigenvalue constraints

```
Step 1 Error 0.4172251847082807 Lambda 128.0
Step 2 Error 0.4093231548475819 Lambda 256.0
Failed to find appropriate step length at iteration 3. Current error 0.4093231548475819. Consider increasing maxlam or lamup.

Eigenvalues:
[ 0.+5.15520237j -0.-5.15520237j  0.+0.49990989j -0.-0.49990989j
  0.+2.51082832j -0.-2.51082832j]
Amplitudes:
[0.21718692 0.21718692 2.00250734 2.00250734 0.99313333 0.99313333]
```

## Summary

It looks like, at least for this particular dataset, applying imaginary eigenvalue constraints makes the solution worse, both in the Matlab and Python versions, regardless of whether my fix is applied or not. My fix makes it better if no eigenvalue constraints are applied, and gives the same result  as the original Matlab version. Note that my fix also gives the same result as the Matlab code when imaginary eigenvalue constraints are applied, which is not the case without my fix.

**Comment by klapo** [comment]
@dsj976 thanks for your hard work. This is a fascinating result. Do you have any intuition or insight why the var pro algorithm might fail when imposing eigenvalue constraints? @sichinaga could we have a conversation about these results and what they mean for eigenvalue constraints generally? Something really weird is going on.

**Comment by dsj976** [comment]
That's great news @klapo! Looking forward to trying it out.

**Comment by klapo** [comment]
I have been algorithm bug hunting for the last few weeks. I want to summarize some of the algorithm failures I've found and specify my preferred solution.

## Effect of the variable projection fix

The variable projection bug fix _does_ result in nicer solution. Specifically, it seems to get out of the local minimum around the initial guess that approximates the exact DMD solution. It does this by increasing the real eigenvalue solution to a (nonsensical) level before stepping it down to more reasonable values. 

## Poor solver states

However, it also results in the solver creating nonsense model states. Some of these states are:
1) Sending the real eigenvalues to such a large value that the exponential growth across the time window goes to infinity, resulting in an SVD convergence error as the solved $\phi$ values become `np.inf`s.
2) Creating multiple conjugate pairs with massive amplitudes (i.e. > 10**10). The conjugate pairs cancel out to a near zero reconstruction.
3) Generally having difficulty fitting data with non-stationarities when any sort of constraint is imposed on the real eigenvalues
4) Generally having difficulty fitting data with non-stationarities when the `svd_rank` is too large (e.g., an ok fit for `svd_rank=6` but a bonkers fit for `svd_rank=10`).

## Current solutions

I have adjusted the variable projection algorithm on a private branch to:
- Force an amplitude limit by imposing that a mode's amplitude cannot be greater than the L2 norm of the data. This addresses problem (1), but revealed problems (2-4).
- Performing the variable projection twice: once with very loose constraints on the real eigenvalue (allow the real eigenvalues to create 10**16 growth or decay) and a second time with whatever constraints the user passes for the real eigenvalues. This solves problem (2) and _some_ of problems (3) and (4). It is unfortunately not robust.

## Proposed solution

Based on a conversation with @dsj976 yesterday, we roughly proposed that we create a flag in the variable projection.  The default would be to do the "true" variable projection with the fix from this pull request as well as my constraints. We would then also have the fall back to do a "fake  variable projection as implemented before, which effectively results in solutions near those from exact DMD. I would document this flag as being a fall back in case the user finds bad fits like I've been encountering.

@sichinaga @mtezzele Would this solution be agreeable? Do you need any other details?

**Pull request merged**